daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/FNFunctions.scala

```diff
@@ -803,6 +807,90 @@ case class FNFloor(recipe: CompiledDPath, argType: NodeInfo.Kind) extends FNOneA
   }
 }
 
+case class DFDLXLeftShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes)  {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+
+    val shiftLong = arg2.getLong
+    val shift = shiftLong.toInt
+    val width = argType.asInstanceOf[PrimNumeric].width.get
+    Assert.invariant(shift >= 0)
+    if(shift>=width)
+      dstate.SDE("Left shift not supported")
+
+    argType match {
+      case NodeInfo.Long => {
+         arg1.getLong << shift
+      }
+      case NodeInfo.Int => {
+         arg1.getInt << shift
+      }
+      case NodeInfo.Short => {
+         (arg1.getShort << shift).toShort
+      }
+      case NodeInfo.Byte => {
+        (arg1.getByte << shift).toByte
+      }
+      case NodeInfo.UnsignedLong =>{
+         arg1.getBigInt.shiftLeft(shift)
+      }
+      case NodeInfo.UnsignedInt => {
+        ULong(arg1.getLong << shift).toLong
```

Comment:
I'm not sure if this is correct. Say the long value was the maximum value of an unsigned int, which is stored in the infoset as a long. That would be `0xFFFFFFFFL`. If we want to shift that left one bit, we get the following:

```scala
scala> passera.unsigned.ULong(0xFFFFFFFFL << 1).toLong
res1: Long = 8589934590
```
In hex, that value is 01 FF FF FF FE, which isn't correct. We have shifted left one and the least signfigcant bit is correctly zero (FE), but there's an extra bit (01) that should have been masked off. I suspect that what we want to do it use the UInt logic for the shift. So we need to first convert our Long value to a UInt, shift the UInt, and then convert it back to a Long since that's what the infoset requires.

So I think we want something like this:
```scala
case NodeInfo.UnsignedInt => (UInt(arg1.getLong.toInt) << 1).toLong
```
I've only tested a couple small edge cases, but that seems to work as exected:
```scala
scala> (passera.unsigned.UInt(0xFFFFFFFFL.toInt) << 1).toLong
res1: Long = 4294967294
scala> (passera.unsigned.UInt(1L.toInt) << 1).toLong
res2: Long = 2
```
I think maybe for the UnsignedLong case, we need to do something slightly different since it's a BigInteger in teh infoset.
```scala
case NodeInfo.UnsignedInt => (ULong(arg1.getBigInteger.longValue) << 1).toBigInt
```
However, the obvious thing above doesn't seem to work for an the max ulong edge case:
```scala
scala> val maxUnsignedLong = new java.math.BigInteger("18446744073709551615")
maxUnsignedLong: java.math.BigInteger = 18446744073709551615
scala> (passera.unsigned.ULong(maxUnsignedLong.longValue) << 1).toBigInt
res1: java.math.BigInteger = 9223372036854775807
```
The result of that shift left should be maxUnsignedLong - 1, but we're instead getting 0x7FFFFFFFFFFFFFFF. So it looks like we're getting a shift right behavior instead of shift left. Not sure what the issue might be...

I think we need something similar for all the shift function as well.

Reply:
@stevedlawrence I have made the changes as per your suggestions.
Can you tell me why case NodeInfo.UnsignedInt=> (arg1.getlong << shift).longValue   does not work.

Reply:
That doesn't work because it can shift outside the range of an UnsignedInt. Note that the valid range of an unsigned int is 0 to 2<sup>32</sup> - 1, or 0xFFFFFFFF in hex, or 4294967295 in decimal.

As an example, say we take that max value of an unsigned int and shift it left 1 bit. In that case, the least significant bit becomes a zero, and the most significant bit is lost since it shifts off the end and outside the width of an unsigned bit. So 0xFFFFFFFF << 1 should result in a value of 0xFFFFFFFE (or 4294967294 in decimal).

Now, say we have this same unsigned int max value, but represent it as a long instead of an unsigned int. The width of a long is 64 bits (compared to the 32 bits of an unsigned int). So if we take our 0xFFFFFFFF as a long and shift it left one, the least signfiicant bit becomes a zero just as before, but now there is addition space for the most significant bit (because widht is 64 instead of 32), and so we don't lose the most significant bit. The result of this is 0x1FFFFFFFE (or 8589934590 in decimal), which is too big that the expected value.

So because we're using a long (which has a larger width than an unsigned it) we can shift to values that fit in a long, but don't fit into an unsigned int.

So when we do `(arg1.getlong << shift).longValue` we are shifting a long, and so we can get values that are too big.

By instead doing `(UInt(arg1.getLong.toInt) << shift).longValue`, we convert that long value to a UInt. When we shift a UInt it will ensure we dont' shift outside of the range of a UInt. After we've done that shift we then convert it back to a long, because that's what the infoset currently requires.